thoth/common/workflows.py

```diff
@@ -0,0 +1,234 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Workflow management for Thoth."""
+
+import logging
+import json
+import random
+import re
+import requests
+import yaml
+
+from pathlib import Path
+
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Mapping
+from typing import Optional
+from typing import Union
+
+from attrdict import AttrDict
+
+from argo.workflows import client
+from argo.workflows import models
+
+from .exceptions import ConfigurationError
+from .exceptions import WorkflowError
+
+from .helpers import to_camel_case
+from .helpers import to_snake_case
+
+from .openshift import OpenShift
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class Workflow(models.V1alpha1Workflow):  # type: ignore
+    """Argo Workflow instance.
+
+    This is a subclass of argo.workflows V1alpha1Workflow model
+    which provides a convenient set of methods to make workflow
+    managemend easier.
+    """
+
+    def __init__(
+        self,
+        api_version: str,
+        kind: str,
+        metadata: models.V1alpha1Metadata,
+        spec: models.V1alpha1WorkflowSpec,
+        status: Optional[models.V1alpha1WorkflowStatus] = None,
+    ):
+        """Initialize Workflow instance."""
+        super().__init__(
+            api_version=api_version, kind=kind, metadata=metadata, spec=spec, status=status or {},
+        )
+
+        if not status:
+            # fixup the status empty dict to prevent serialization issues
+            self.status = None
+
+        self.__validated = False
+
+    @property
+    def name(self) -> Union[str, None]:
+        """Get Workflow name."""
+        name: Union[str, None] = getattr(self.metadata, "name", None)
+        return name
+
+    @property
+    def id(self) -> str:
+        """Get Workflow ID."""
+        digest: str = abs(self.__hash__())
+        return f"workflow{'-' + self.name if self.name else ''}-{digest}"
+
+    @property
+    def validated(self) -> bool:
+        """Return whether this workflow has been validated."""
+        return self.__validated
+
+    def __eq__(self, other: Any) -> Any:
+        """Compare workflows for equality."""
+        return self.id == other.id
+
+    def __hash__(self) -> Any:
+        """Compute hash of this Workflow."""
+        return self.to_str().__hash__()
+
+    @classmethod
+    def from_file(cls, fp: Union[str, Path], fmt: str = "yaml") -> "Workflow":
+        """Create a Workflow from a file."""
+        fmt = fmt.lower()
+
+        wf_path = Path(fp)
+
+        wf: Dict[str, Any]
+        if fmt == "yaml":
+            wf = yaml.safe_load(wf_path.read_text())
```

Comment:
JSON is a subset of YAML, you can discard `fmt` argument and use just `yaml.safe_load`:

```python
>>> import yaml
>>> yaml.safe_load('{"foo": 1, "bar": ["baz"]}')
{'foo': 1, 'bar': ['baz']}
```

Reply:
That's neat!